infrastructure/src/main/java/org/corfudb/infrastructure/DynamicTriggerPolicy.java

```diff
@@ -0,0 +1,118 @@
+        this.lastAddressSpaceSizeAfterTrim = getCurrentAddressSpaceSize();
+    }
+
+    private long getCurrentAddressSpaceSize() {
+        long currentAddressSpaceSize = 0;
+        final int maxRetries = 10;
+        for (int retry = 0; retry < maxRetries; retry++) {
+            try {
+                currentAddressSpaceSize = corfuRuntime.getAddressSpaceView().getLogTail()
+                        - corfuRuntime.getAddressSpaceView().getTrimMark().getSequence();
+            } catch (Exception e) {
+                syslog.warn("getCurrentAddressSpaceSize hit exception {}. StackTrace {}", e, e.getStackTrace());
+            }
+        }
+        return currentAddressSpaceSize;
+    }
+
+    private boolean shouldForceTrigger() {
+        CorfuStore corfuStore = new CorfuStore(corfuRuntime);
+        try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+            RpcCommon.TokenMsg upgradeToken = (RpcCommon.TokenMsg) txn.getRecord(DistributedCompactor.CHECKPOINT,
+                    DistributedCompactor.UPGRADE_KEY).getPayload();
+            txn.commit();
+            if (upgradeToken != null) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * 1. if ((currentTime - lastCompactionCycleStart) > minTimeBetweenCompactionStarts)
+     *        if (lastAddressSpaceSizeOnTrim == 0)
+     *            return true // no record of previous trim & safe trim period elapsed
+     *            ... once trim happens, we record the lastAddressSpaceSizeOnTrim
+     *            ... once checkpoint starts we record the lastCompactionCycleStartTS
+     * 2. if ((currentTime - lastCompactionCycleStart) > maxTimeBetweenCompactionStarts)
+     *
+     * @param interval - ignored for now
+     * @return true if compaction cycle should run, false otherwise
+     */
+    @Override
+    public boolean shouldTrigger(long interval) {
+
+        if (shouldForceTrigger()) {
+            syslog.info("Force triggering compaction");
+            return true;
+        }
+
+        final long currentTime = System.currentTimeMillis();
+        final long timeSinceLastCycleMillis = currentTime - lastCompactionCycleStartTS;
+        if (timeSinceLastCycleMillis > interval && lastAddressSpaceSizeAfterTrim == 0) {
+            // no record of previous trim & safe trim period elapsed when lastAddressSpaceSizeAfterTrim is 0
+            syslog.info("DynamicTriggerPolicy: Trigger as elapsedTime {} > safeTrimPeriod {}",
+                    TimeUnit.MILLISECONDS.toSeconds(timeSinceLastCycleMillis),
+                    TimeUnit.MILLISECONDS.toSeconds(interval));
+            return true;
+        }
+
+        if (timeSinceLastCycleMillis > interval*2) {
+            syslog.info("DynamicTriggerPolicy: Trigger as elapsedTime {} > maxTimeToChkpt {}",
+                    TimeUnit.MILLISECONDS.toMinutes(timeSinceLastCycleMillis),
+                    TimeUnit.MILLISECONDS.toMinutes(interval*2));
```

Comment:
code reformat.

And for all the code, please make sure that code formatted properly